src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,351 @@
+                                    //    else { Use(x); }
+                                    //
+                                    return false;
+                                }
+
+                                if (dataFlow.AlwaysAssigned.Contains(_localSymbol))
+                                {
+                                    // If the variable is always assigned here, we don't need to check
+                                    // subsequent statements as it's definitely assigned afterwards.
+                                    // For example:
+                                    //
+                                    //     if (o is string x) { }
+                                    //     else { x = null; }
+                                    //
+                                    return true;
+                                }
+                            }
+
+                            if (!defAssignedWhenTrue &&
+                                !_semanticModel.AnalyzeControlFlow(ifStatement.Statement).EndPointIsReachable)
+                            {
+                                // Access before assignment here is only valid if we have a negative
+                                // pattern-matching in an if-statement with an unreachable endpoint.
+                                // For example:
+                                //
+                                //      if (!(o is string x)) {
+                                //        return;
+                                //      }
+                                //
+                                //      // The 'return' statement above ensures x is definitely assigned here
+                                //      Console.WriteLine(x);
+                                //
+                                return true;
+                            }
+
+                            return CheckStatement(ifStatement);
+                    }
+
+                    switch (current)
+                    {
+                        case ExpressionSyntax expression:
+                            // If we reached here, it means we have a sub-expression that
+                            // does not garantee definite assignment. We should make sure that
+                            // the pattern variable is not used outside of the expression boundaries.
+                            return CheckExpression(expression);
+
+                        case StatementSyntax statement:
+                            // If we reached here, it means that the null-check is appeared in
+                            // a statement. In that case, the variable would be actually in the
+                            // scope in subsequent statements, but not definitely assigned.
+                            // Therefore, we should ensure that there is no use before assignment.
+                            return CheckStatement(statement);
+
+                        default:
```

Comment:
does this walking up work if you're in a lamdba with an expression body?

Reply:
good catch. probably not

Reply:
turns out it does. `=>` in lambda does not have a separate node, so we have an Expression here.